### DISCUSSION TITLE
Change the 24H download stats

### DISCUSSION THREAD

**Discussion by serkor1** [flaming-codes/crane-app#170] [General]
At the time of writing packages like `tidyverse`, `data.table` and `Rcpp` shows 0 downloads for the last 24 hours. This is clearly wrong.

The error comes, I believe, from how CRAN updates the package downloads. See the following for an elaboration,

## Last-day download counts

```R
cranlogs::cran_downloads(
  packages = 'tidyverse',
  when = 'last-day'
)
```
```R
        date count   package
1 2023-12-28 30867 tidyverse
```

Using `last-day` gives the latest published downloads, which corresponds to the last (available) 24 hour downloads.

## Date-specific download counts

```R
cranlogs::cran_downloads(
  packages = 'tidyverse',
  from = as.character(
    Sys.Date() - 1
  )
)
```
```R
        date count   package
1 2023-12-29     0 tidyverse
2 2023-12-30     0 tidyverse
```

Using `Sys.Date() - 1`, returns the last, or last two days depending on what time you make the call, dates and will return it as 0 if you call it before the numbers are published, or we pass into the next day.

-----

I am not familiar with the language you are using, so I couldn't locate the error. But my guess is that your API calls uses date-specific calls, which doesn't return any sensible value until CRAN releases the actual numbers. But when time passes midnight (I guess) you are back to the date-specific issue which returns 0 downloads.

> At the time of writing, I can't for the best of me find the exact time and date of when CRAN publishes these numbers. But it clearly plays a role on how you pull the numbers to your PWA.

**Comment by flaming-codes** [comment]
Hi @serkor1, thanks for taking the time and the detailed write-up.

I agree with your point that the current `Last 24 hours` in CRAN/E is misleading. 

## tl;dr
The official logs-API (consumed via the JSON-endpoint in CRAN/E) can't deliver fresh data, earliest data is from "last day", I assume relative to UTC. CRAN/E current doesn't want to show stale data, so `0` is shown until the logs-API provides the latest data (from last day, though).

### Immediate action
I'll update the wording + API-call to always use `Yesterday` and not `Last 24 hours`, as this data simply doesn't exist. This will fix the issue of count `0`.


## More context 

<summary>
<details>
<br/>
After looking a little bit more into it, I understand that 

- Logs-DB
- Trigger/Logic to update Logs-DB (likely part of package manager service)

are closed source, therefore we have no ability to hook into this system.


Apparently, the Logs-DB (i.e. downloads + trends metadata) is a simple Postgres-DB. The [SQL-code](https://github.com/r-hub/cranlogs.app/blob/master/db/create.sh) in `cranlogs.app`-repo for creating the initial schema tells me this. Interestingly, it looks like the DB is only using built-in Postgres features _and not the Timescale-DB-extension_, which is purpose-built for time series data. `TimescaleDB` has some not-so-nice licensing if you're using the extension as part of a hosted service (e.g. Supabase), but if you're self-hosting your Postgres-instance, all features of `TimescaleDB` are available, including their auto-updating variant of materialised views for time-bucket sliced data.

In other words, if the referenced script is showing us the actual schema of the DB, CRAN is not using Postgres to its full extent, not even close TBH. A refactor of the DB would be required to use `timescaledb`-extension w/ their `time-bucket`-function and `hyper tables`, so that apart from the current set of slices, we also have _hourly_ time-buckets. This would allow for true time-series retrieval.

As of writing, CRAN has 21,520 packages. Assuming 20% of all packages are used within any given hour on average then 4,304 entries would be aggregated for each hour. The schema for the "daily"-table is currently minimal:

- date: (DATE) 4 Bytes
- package: (VARCHAR) max 50 Bytes, but Postgres makes the decision
- count: (BIGINT) 8 Bytes
- sum (Aggregated as part of materialised view, same as count) 8 Bytes

So, assuming one package/hour has 70 Bytes, we would be at **301.28 KB per hour**, thus having ~7.2 MB/day. This would yield 216 MB per month for a time-bucket with an hourly-scope. This size can be drastically reduced by only projecting e.g. the most 100 downloaded packages (as it's done now in the referenced script), so the size would be _minimal_.

## Conclusion

I would really like to set up the time series database for CRAN 😄 sadly that's not possible, as the real data/access to it is closed source. The current implementation of the `cranlogs` seems to be quite simplistic w/o the relevant `timeseriesdb`-extension to have a granular temporal filter for aggregations that can scale.

Thanks again for taking a look into it @serkor1 👍 

</details>
<summary>
